### PR TITLE
Feature 377

### PR DESCRIPTION
**Pull request by FFroehlich**
Improve ExpData interface and move some ForwardProblem logic to Model, partially adresses #377.

### CODE REVIEW

File: src/edata.cpp [hunk 1]
```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
...
+        return nullptr;
+}
 
-void ExpData::setObservedDataStdDev(const double *observedDataStdDev) {
-    for (int imy = 0; imy < nytrue * nt; ++imy) {
-        sigmay.at(imy) = static_cast<const realtype>(observedDataStdDev[imy]);
+void ExpData::setObservedDataStdDev(const std::vector<realtype> &observedDataStdDev) {
+    if (observedDataStdDev.size() == (unsigned) nt()*nytrue) {
+        this->observedDataStdDev = observedDataStdDev;
+    } else {
```
Comment (dweindl) [inline src/edata.cpp]: Avoid unnecessary indentation, use `else if`

File: src/edata.cpp [hunk 2]
```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
...
+        if (observedDataStdDev.size() == 0) {
```
Comment (dweindl) [inline src/edata.cpp]: .empty()

File: src/edata.cpp [hunk 3]
```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
...
+void ExpData::setObservedDataStdDev(const realtype StdDev, int iy) {
+    for (int it = 0; it < nt(); ++it)
+        observedDataStdDev.at(iy + it*nytrue) = StdDev;
+}
+    
+bool ExpData::isSetObservedDataStdDev(int it, int iy) const {
+    if (!observedDataStdDev.empty()) {
+        return !isNaN(observedDataStdDev.at(it * nytrue + iy));
+    } else {
+        return false;
```
Comment (dweindl) [inline src/edata.cpp]: `return !observedDataStdDev.empty() && !isNaN(observedDataStdDev.at(it * nytrue + iy));`

File: src/edata.cpp [hunk 4]
```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
...
+    return observedDataStdDev;
+}
 
-void ExpData::setObservedEvents(const double *observedEvents) {
-    for (int imz = 0; imz < nztrue * nmaxevent; ++imz) {
-        mz.at(imz) = static_cast<const realtype>(observedEvents[imz]);
+const realtype *ExpData::getObservedDataStdDevPtr(int it) const {
+    if (!observedDataStdDev.empty()) {
+        return &observedDataStdDev.at(it*nytrue);
+    } else {
```
Comment (dweindl) [inline src/edata.cpp]: Remove `else`, just `return nullptr`

File: src/edata.cpp [hunk 5]
```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
...
+        return nullptr;
     }
 }
+    
+void ExpData::setObservedEvents(const std::vector<realtype> &observedEvents) {
+    if (observedEvents.size() == (unsigned) nmaxevent*nztrue) {
+        this->observedEvents = observedEvents;
+    } else {
```
Comment (dweindl) [inline src/edata.cpp]: Use `else if`

File: src/edata.cpp [hunk 6]
```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
...
+        if (observedEvents.size() == 0) {
+            this->observedEvents.clear();
+        } else {
+            throw AmiException("Input observedEvents did not match dimensions nmaxevent (%i) x nztrue (%i), was %i", nmaxevent, nztrue, observedEvents.size());
+        }
+    }
+}
+    
+void ExpData::setObservedEvents(const std::vector<realtype> &observedEvents, int iz) {
+    if (observedEvents.size() == (unsigned) nmaxevent) {
```
Comment (dweindl) [inline src/edata.cpp]: Reduce nesting for readability
```
if (observedEvents.size() != (unsigned) nmaxevent) {
  throw AmiException("Input observedEvents did not match dimensions nmaxevent (%i), was %i", nmaxevent, observedEvents.size());
} 

for (int ie = 0; ie < nmaxevent; ++ie)
    this->observedEvents.at(iz + ie*nztrue) = observedEvents.at(iz); 
```

File: src/edata.cpp [hunk 7]
```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
...
+        for (int ie = 0; ie < nmaxevent; ++ie)
+            this->observedEvents.at(iz + ie*nztrue) = observedEvents.at(iz);
+    } else {
+        throw AmiException("Input observedEvents did not match dimensions nmaxevent (%i), was %i", nmaxevent, observedEvents.size());
+    }
+}
+    
+bool ExpData::isSetObservedEvents(int ie, int iz) const {
+    if (!observedEvents.size()) {
```
Comment (dweindl) [inline src/edata.cpp]: return !observedEvents.empty() && !isNaN(observedEvents.at(ie * nztrue + iz));

File: src/edata.cpp [hunk 8]
```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
...
+    return observedEvents;
+}
 
-void ExpData::setObservedEventsStdDev(const double *observedEventsStdDev) {
-    for (int imz = 0; imz < nztrue * nmaxevent; ++imz) {
-        sigmaz.at(imz) = static_cast<const realtype>(observedEventsStdDev[imz]);
+const realtype *ExpData::getObservedEventsPtr(int ie) const {
+    if (!observedEvents.empty())
+        return &observedEvents.at(ie*nztrue);
+    else
```
Comment (dweindl) [inline src/edata.cpp]: remove superfluous `else`

File: src/edata.cpp [hunk 9]
```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
...
+        return nullptr;
+}
+    
+void ExpData::setObservedEventsStdDev(const std::vector<realtype> &observedEventsStdDev) {
+    if (observedEventsStdDev.size() == (unsigned) nmaxevent*nztrue) {
+        this->observedEventsStdDev = observedEventsStdDev;
+    } else {
```
Comment (dweindl) [inline src/edata.cpp]: else if

File: src/edata.cpp [hunk 10]
```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
...
+        }
+    }
+}
+    
+void ExpData::setObservedEventsStdDev(const realtype StdDev) {
+    std::fill(observedEventsStdDev.begin() ,observedEventsStdDev.end(), StdDev);
+}
+
+void ExpData::setObservedEventsStdDev(const std::vector<realtype> &observedEventsStdDev, int iz) {
+    if (observedEventsStdDev.size() == (unsigned) nmaxevent) {
```
Comment (dweindl) [inline src/edata.cpp]: See above

File: src/edata.cpp [hunk 11]
```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
...
+    }
+}
+
+void ExpData::setObservedEventsStdDev(const realtype StdDev, int iz) {
+    for (int ie = 0; ie < nmaxevent; ++ie)
+        observedEventsStdDev.at(iz + ie*nztrue) = StdDev;
+}
+    
+bool ExpData::isSetObservedEventsStdDev(int ie, int iz) const {
+    if (!observedEventsStdDev.empty()) {
```
Comment (dweindl) [inline src/edata.cpp]: return !observedEventsStdDev.empty() && !isNaN(observedEventsStdDev.at(ie * nztrue + iz)):

File: src/edata.cpp [hunk 12]
```diff
@@ -89,40 +75,209 @@ ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::ve
...
+    } else {
+        return false;
     }
 }
 
+std::vector<realtype> ExpData::getObservedEventsStdDev() const {
+    return observedEventsStdDev;
+}
+
+const realtype *ExpData::getObservedEventsStdDevPtr(int ie) const {
```
Comment (dweindl) [inline src/edata.cpp]: see above

File: src/forwardproblem.cpp [hunk 13]
```diff
@@ -434,21 +431,9 @@ void ForwardProblem::prepEventSensis(int ie) {
...
-                model->dsigmazdp[iz + model->nz * ip] = 0;
-            }
-            model->sigmaz[iz] =
-                    edata->sigmaz[nroots.at(ie) * model->nztrue + iz];
-        }
-        rdata->sigmaz[nroots.at(ie) * model->nz + iz] = model->sigmaz[iz];
-        for (int ip = 0; ip < model->nplist(); ip++) {
-            rdata->ssigmaz[(nroots.at(ie)*model->np()+ip) * model->nz + iz] =
-                    model->dsigmazdp[iz + model->nz * ip];
-        }
```
Comment (dweindl) [inline src/forwardproblem.cpp]: :+1:

File: src/forwardproblem.cpp [hunk 14]
```diff
@@ -564,27 +549,8 @@ void ForwardProblem::getDataSensisFSA(int it) {
...
-                }
-            }
-            for (int ip = 0; ip < model->nplist(); ip++) {
-                rdata->ssigmay[(it * model->nplist() + ip) * model->ny + iy] =
-                        model->dsigmaydp[ip * model->ny + iy];
-            }
-        }
-    } else {
-        std::fill_n(&rdata->ssigmay[it * model->nplist() * model->ny], model->nytrue * model->nplist(), 0.0);
-    }
```
Comment (dweindl) [inline src/forwardproblem.cpp]: :+1:

File: src/model.cpp [hunk 15]
```diff
@@ -897,10 +902,30 @@ void Model::fsigmaz(const realtype t, const int ie, const int *nroots,
 /** Sensitivity of standard deviation of events measurements w.r.t. model parameters p
      * @param t current timepoint
      */
-void Model::fdsigmazdp(const realtype t) {
+void Model::fdsigmazdp(const realtype t, const int ie, const int *nroots, ReturnData *rdata, const ExpData *edata) {
     std::fill(dsigmazdp.begin(),dsigmazdp.end(),0.0);
     for(int ip = 0; (unsigned)ip < plist_.size(); ip++)
-        fdsigmazdp(&dsigmazdp.at(ip*nz),t, unscaledParameters.data(),fixedParameters.data(),plist_.at(ip));
+        // get dsigmazdp slice (nz) for current event and parameter
+        fdsigmazdp(&dsigmazdp.at(ip*nz),
```
Comment (dweindl) [inline src/model.cpp]: please use braces for multiline for-body

File: src/model.cpp [hunk 16]
```diff
@@ -897,10 +902,30 @@ void Model::fsigmaz(const realtype t, const int ie, const int *nroots,
...
+                   t,
+                   unscaledParameters.data(),
+                   fixedParameters.data(),
+                   plist_.at(ip));
+    
+    // sigmas in edata override model-sigma -> for those sigmas, set dsigmazdp to zero
+    if(edata) {
+        for (int iz = 0; iz < nztrue; iz++) {
+            if (z2event.at(iz) - 1 == ie) {
```
Comment (dweindl) [inline src/model.cpp]: Reduce nesting: 
`if (z2event.at(iz) - 1 != ie) continue`

File: src/edata.cpp [hunk 17]
```diff
@@ -10,49 +10,37 @@
...
-    : ExpData(nytrue, nztrue, nt, nmaxevent,
-              std::vector<realtype>(nt * nytrue),
-              std::vector<realtype>(nt * nytrue),
-              std::vector<realtype>(nmaxevent * nztrue),
-              std::vector<realtype>(nmaxevent * nztrue))
+ExpData::ExpData() : nytrue(0), nztrue(0), nmaxevent(0) {}
 
+ExpData::ExpData(int nytrue, int nztrue, int nmaxevent,
+                 std::vector<realtype> ts)
+    : nytrue(nytrue), nztrue(nztrue), nmaxevent(nmaxevent)
```
[the remark this answers is not in the document]
Comment (FFroehlich) [reply src/edata.cpp]: no, `setTimepoints` also resizes `observedData` and `observedDataStdDev` and fills them with `Nan`.

File: include/amici/edata.h [hunk 18]
```diff
@@ -95,66 +87,248 @@ class ExpData {
...
     /** maximal number of event occurences */
     const int nmaxevent;
-
+    
     /** condition-specific parameters of size Model::nk() or empty */
     std::vector<realtype> fixedParameters;
     /** condition-specific parameters for pre-equilibration of size Model::nk() or empty.
-      * Overrides Solver::newton_preeq */
+     * Overrides Solver::newton_preeq */
     std::vector<realtype> fixedParametersPreequilibration;
```
[the remark this answers is not in the document]
Comment (FFroehlich) [reply include/amici/edata.h]: we currently cannot perform dimension checks on those two as we do not know about nk, so there is no point in making them private but having to write additional getter/setter routines.

File: src/edata.cpp [hunk 19]
```diff
@@ -61,15 +49,13 @@ ExpData::ExpData(ReturnData const& rdata, realtype sigma_y, realtype sigma_z)
 }
     
 ExpData::ExpData(ReturnData const& rdata, std::vector<realtype> sigma_y, std::vector<realtype> sigma_z)
-    : ExpData(rdata.nytrue, rdata.nztrue, rdata.nt, rdata.nmaxevent)
+    : ExpData(rdata.nytrue, rdata.nztrue, rdata.nmaxevent, rdata.ts)
 {
-    if (sigma_y.size() != nytrue)
+    if (sigma_y.size() != (unsigned) nytrue)
```
[the remark this answers is not in the document]
Comment (FFroehlich) [reply src/edata.cpp]: I do not think assigning the whole std matrix will every get used, this is only per observable (made this more explicit in docstring)

File: src/edata.cpp (hunk 1 above)
Comment (FFroehlich) [reply src/edata.cpp]: `else if` would mean that `observedData.size() == (unsigned) nt()*nytrue` will not be checked before 
```
        if (observedData.empty()) {
            this->observedData.clear();
        } else {
            throw AmiException("Input observedData did not match dimensions nt (%i) x nytrue (%i), was %i", nt(), nytrue, observedData.size());
        }
```
is executed. Hence we would first call `this->observedData = std::move(observedData);` and then throw `throw AmiException("Input observedData did not match dimensions nt (%i) x nytrue (%i), was %i", nt(), nytrue, observedData.size());` in most scenarios, which is definitely not what we want.

File: src/edata.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (dweindl) [reply src/edata.cpp]: Okay. People might stumble over those cleared data when resetting timepoints...

File: src/edata.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (dweindl) [reply src/edata.cpp]: My code currently does that... :)

File: src/edata.cpp (hunk 3 above)
Comment (FFroehlich) [reply src/edata.cpp]: calling `!observedDataStdDev.empty()` first that we do not perform out of bound memory access on `observedDataStdDev`.

File: src/edata.cpp (hunk 5 above)
Comment (FFroehlich) [reply src/edata.cpp]: see above

File: src/edata.cpp (hunk 1 above)
Comment (dweindl) [reply src/edata.cpp]: Not sure I get that.
This
```
if(a) {
    A
} else {
    if(b) {
        B
    } else {
        C
    }
}
```
is identical with
```
if(a) {
    A
} else if(b) {
    B
} else {
    C
}
```
except for verbosity...

File: src/edata.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (FFroehlich) [reply src/edata.cpp]: okay lets remove the clear then and only call resize

File: src/edata.cpp (hunk 1 above)
Comment (FFroehlich) [reply src/edata.cpp]: oh, misunderstood a stackoverflow post then :(

File: src/edata.cpp (hunk 3 above)
Comment (FFroehlich) [reply src/edata.cpp]: oh, neat

**Pull request merged**